storage/innobase/os/os0file.cc

```diff
@@ -4769,6 +4796,75 @@ readahead requests. */
 #endif /* __WIN__ */
 }
 
+/** Submit buffered AIO requests on the given segment to the kernel
+(low level function).
+@param acquire_mutex specifies whether to lock array mutex
+*/
+static
+void
+os_aio_dispatch_read_array_submit_low(bool acquire_mutex MY_ATTRIBUTE((unused)))
+{
+#if defined(LINUX_NATIVE_AIO)
+	os_aio_array_t*	array = os_aio_read_array;
+	ulint		total_submitted = 0;
+	if (!srv_use_native_aio) {
+		return;
+	}
+	if (acquire_mutex)
+		os_mutex_enter(array->mutex);
+	/* Submit aio requests buffered on all segments. */
+	for (ulint i = 0; i < array->n_segments; i++) {
+		int	count = array->count[i];
+		if (count > 0) {
+			ulint	iocb_index = i * array->n_slots
+					     / array->n_segments;
+			int	submitted = io_submit(
+					array->aio_ctx[i], count,
+					&(array->pending[iocb_index]));
+			if (submitted == count) {
+				total_submitted += submitted;
+			} else {
+				/* io_submit returns number of successfully
+				queued requests or -errno. */
+				if (submitted >= 0) {
+					ib_logf(IB_LOG_LEVEL_ERROR,
```

Comment:
Is it possible for io_submit to return partial success value < count? Looking at its interface, I don't see how, either -errno, either complete success, leaving no means to restart a partial submission. If this is correct, add a DBUG_ASSERT(0)?

Reply:
OK, ideally this has to be loop. Please double-check ret == 0 case, from my reading, it's impossible unless submitted array was also zero-length, this could be asserted I guess.

Reply:
Yes, you are right. From the comments and the source code itself, it looks like `ret == 0` only when `nr == 0`. If `io_submit()` fails at the very first `iocb` it returns actual `errno` (propagated from `io_submit_one()`)
https://github.com/torvalds/linux/blob/master/fs/aio.c#L1700